### PR TITLE
EVA-839 assembly check

### PR DESCRIPTION
**Pull request by pabarcgar**

### CODE REVIEW

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                                                                             "NT_455997.1:g.1330026_1330028delTAT",
+                                                                             1330026L, 1330028L, Orientation.REVERSE,
+                                                                             null, "1062064");
+
+        assertNull(assemblyChecker.process(deletionWithInvalidReference));
+    }
+
+    
+    @Test
+    public void validMnv() throws Exception {
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessorTest.java]: I think this is considered an INDEL, an MNV would be something like TC -> AG (same length).
Also applies to method `invalidMnv`.

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                                                                        "NT_455997.1:g.1340260_1340261delTTinsA",
+                                                                        1340260L, 1340261L, Orientation.FORWARD, null,
+                                                                        "1055116");
+
+        assertNull(assemblyChecker.process(mnvWithInvalidReference));
+    }
+
+    @Test
+    public void validSnpWithNoChromosomeCoordinates() throws Exception {
+        // "RefseqToGenbankMappingProcessor" replaces the Refseq contig for an Esembl one in the "contig" field of the
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessorTest.java]: It should be GenBank instead of Ensembl. Also applies to `invalidSnpWithNoChromosomeCoordinates`.

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Test
+    public void getLastNucleotideOfContig() throws Exception {
+        assertEquals("G", reader.getSequence("22", 4729743, 4729743));
+    }
+
+    @Test
+    public void getSequence() throws Exception {
+        // this sequence is splitted between three lines in the fasta file
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java]: "split" instead of "splitted"

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        assertEquals("GTTTCAAGTGGTTGTGACCCCCGCTGCACAGTCAGTTGGGTTAGGGTTAGGGTTAGGGTCAGTCACAGTCAGTTGTCAGACTGGTGTTTA",
+                     reader.getSequence("22", 59986, 60075));
+    }
+
+    @Test
+    public void endMustBeGreaterOrEqualsThanStart() throws Exception {
+        thrown.expect(ReadSequenceException.class);
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java]: If the arguments provided are not valid, the ideal exception is `IllegalArgumentException`.

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        reader.getSequence("22", 1000, 999);
+    }
+
+    @Test
+    public void justPositiveCoordinatesAreAllowed() throws Exception {
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java]: "only" instead of "just"

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        thrown.expect(ReadSequenceException.class);
+        reader.getSequence("22", -1, 5);
+    }
+
+    @Test
+    public void coordinatesExcedingEndOfChromosomeAreNotAllowed() throws Exception {
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java]: "greater than" is more idiomatic than "exceeding"

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/configuration/processors/AssemblyCheckFilterProcessorConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Bean
+    AssemblyCheckFilterProcessor assemblyCheckFilterProcessor(Parameters parameters) {
+        Path referenceFastaFile = Paths.get(parameters.getReferenceFastaFile());
+        fastaSequenceReader = new FastaSequenceReader(referenceFastaFile);
+        return new AssemblyCheckFilterProcessor(fastaSequenceReader);
+    }
+
+    @Bean(FASTA_SEQUENCE_READER_CLOSER)
+    StepListenerSupport fastaSequenceReaderCloser() {
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/configuration/processors/AssemblyCheckFilterProcessorConfiguration.java]: Ok, @jmmut told me that this listener was just for logging, so I created a different one. If he does not mind, I will move this to that listener.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                                                                       region.getEnd());
+                if (referenceAllele.equals(sequenceInAssembly)) {
+                    return subSnpCoreFields;
+                } else {
+                    logger.warn("Variant filtered out because it the reference allele is not correct: {}.\n" +
+                                        "{} expected in {}, {} found",
+                                subSnpCoreFields, sequenceInAssembly, region, referenceAllele);
+                    return null;
+                }
+            } catch (ReadSequenceException e) {
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java]: This can be thrown also if end is greater than start, or if the coordinates are outside the chromosome limits. If we throw the exception and a variant is stored in dbSNP with incorrect coordinates, we are not going to be able to continue the execution for that batch unless we change the code. I'm going to try to separate this in different exceptions, so we stop the execution just if the fasta read cannot be read.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/configuration/processors/AssemblyCheckFilterProcessorConfiguration.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (cyenyxe) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/configuration/processors/AssemblyCheckFilterProcessorConfiguration.java]: I would still rename it, but the code structure is fine.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (cyenyxe) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java]: Yes, in a comment below I suggest using `IllegalArgumentException` in the case of illegal coordinates being provided.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/SequenceReader.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ * limitations under the License.
+ */
+package uk.ac.ebi.eva.dbsnpimporter.sequence;
+
+import java.util.NoSuchElementException;
+
+/**
+ * Interface for readers than returns nucleotide subsequences of a given sequence source (like a fasta file or a WS)
+ */
+public interface SequenceReader {
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/SequenceReader.java]: I was keeping it just in case the cache annotations could be just applied over interfaces (I think they can be applied over classes as well but with a more complex configuration). Like it seems we are not going to use the cache yet, I will delete this interface

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java [hunk 10]
```diff
@@ -21,49 +21,66 @@
...
-                }
-            } catch (ReadSequenceException e) {
-                logger.warn(
-                        "Variant filtered out because the region {} cannot be retrieved from the reference sequence " +
-                                "file: {}\n{}",
-                        region, subSnpCoreFields, e.getMessage());
-                return null;
-            }
+            return null;
+
```
Comment (cyenyxe) [inline dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java]: Remove blank line.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java [hunk 11]
```diff
@@ -21,49 +21,66 @@
...
         }
     }
 
     private boolean isEmpty(String referenceAllele) {
         return referenceAllele == null || referenceAllele.trim().length() == 0 || referenceAllele.equals("-");
     }
+
+    private boolean referenceAlleleIsCorrect(String referenceAllele,
+                                             SubSnpCoreFields subSnpCoreFields) throws ReadSequenceException {
+
```
Comment (cyenyxe) [inline dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java]: Remove blank line.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java [hunk 12]
```diff
@@ -21,49 +21,66 @@
...
+        Region region = subSnpCoreFields.getVariantCoordinates();
+        try {
+            String sequenceInAssembly = fastaReader.getSequence(region.getChromosome(), region.getStart(),
+                                                                region.getEnd());
+            if (referenceAllele.equals(sequenceInAssembly)) {
+                return true;
+            } else {
+                logger.warn(
+                        "Variant filtered out because it the reference allele does not match the reference sequence: " +
```
Comment (cyenyxe) [inline dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/jobs/steps/processors/AssemblyCheckFilterProcessor.java]: "because the reference allele" (there is an extra "it")

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java [hunk 13]
```diff
@@ -16,29 +16,42 @@
...
 
     private final ReferenceSequenceFile fastaSequenceFile;
 
+    private final SAMSequenceDictionary sequenceDictionary;
+
     public FastaSequenceReader(Path fastaFile) {
         fastaSequenceFile = ReferenceSequenceFileFactory.getReferenceSequenceFile(fastaFile, true);
+        sequenceDictionary = fastaSequenceFile.getSequenceDictionary();
+        if (sequenceDictionary == null || !fastaSequenceFile.isIndexed()) {
+            throw new IllegalArgumentException("A sequence dictionary file and a Fasta index file are required");
```
Comment (cyenyxe) [inline dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java]: Isn't the sequence dictionary file "embedded" in the FASTA file? If that is the case I would suggest rewording as:

```
A FASTA file containing a sequence dictionary and a FASTA index file are required
```

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java [hunk 14]
```diff
@@ -47,7 +60,27 @@ public String getSequence(String contig, long start, long end) throws ReadSequen
         }
     }
 
-    @Override
+    private void checkArguments(String contig, long start, long end) throws IllegalArgumentException {
+        if (end < start) {
+            throw new IllegalArgumentException("'end' must be greater or equal than 'start'");
```
Comment (cyenyxe) [inline dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java]: "equal or greater than"

File: dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java [hunk 15]
```diff
@@ -62,25 +62,37 @@ public void getSequence() throws Exception {
...
+    public void fastaWithNoDictionary() throws Exception {
+        thrown.expect(IllegalArgumentException.class);
+        new FastaSequenceReader(Paths.get("src/test/resources/fastaWithNoDictionary.fa"));
+    }
+
+    @Test
+    public void fastaWithNoIndex() throws Exception {
+        thrown.expect(IllegalArgumentException.class);
+        new FastaSequenceReader(Paths.get("src/test/resources/fastaWithNoIndex.fa"));
+    }
```
Comment (cyenyxe) [inline dbsnp-importer/src/test/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReaderTest.java]: Maybe another test for FASTA file not found?

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/parameters/Parameters.java [hunk 16]
```diff
@@ -137,6 +147,8 @@ public String toString() {
                 ", batchId=" + batchId +
                 ", variantsCollection='" + variantsCollection + '\'' +
                 ", chunkSize=" + chunkSize +
+                ", contigMappingUrl='" + contigMappingUrl + '\'' +
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/parameters/Parameters.java]: No worries, we will have to rebase the second PR and fix the conflict there.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    private void checkArguments(String contig, long start, long end) throws IllegalArgumentException {
+        if (end < start) {
+            throw new IllegalArgumentException("'end' must be greater or equal than 'start'");
+        } else if (start < 1) {
+            throw new IllegalArgumentException("'start' and 'end' must be positive numbers");
+        } else if (sequenceDictionary.getSequence(contig) == null) {
+            throw new IllegalArgumentException("Sequence " + contig + " not found in reference fasta file");
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/FastaSequenceReader.java]: This processor is being executed after the refseq to genbank map, so we should find GeneBank contigs in the variants. If we download the fasta files from Ensembl, I think that it will be OK. For the species not supported by Ensembl, I think they are not going to be imported into the data warehouse, so it shouldn' t matter.

File: dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/ReadSequenceException.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package uk.ac.ebi.eva.dbsnpimporter.sequence;
+
+public class ReadSequenceException extends Exception {
+
+    ReadSequenceException(String message) {
+        super(message);
+    }
+
+    ReadSequenceException(RuntimeException e) {
```
[the remark this answers is not in the document]
Comment (pabarcgar) [reply dbsnp-importer/src/main/java/uk/ac/ebi/eva/dbsnpimporter/sequence/ReadSequenceException.java]: Removing ReadSequenceException class.
